clippy_lints/src/loops/mod.rs

```diff
@@ -0,0 +1,669 @@
+            // this check is not at the top of the function
+            // since higher::for_loop expressions are marked as expansions
+            if body.span.from_expansion() {
+                return;
+            }
+            check_for_loop(cx, pat, arg, body, expr, span);
+        }
+
+        // we don't want to check expanded macros
+        if expr.span.from_expansion() {
+            return;
+        }
+
+        // check for never_loop
+        never_loop::check_never_loop(cx, expr);
+
+        // check for `loop { if let {} else break }` that could be `while let`
+        // (also matches an explicit "match" instead of "if let")
+        // (even if the "match" or "if let" is used for declaration)
+        if let ExprKind::Loop(ref block, _, LoopSource::Loop, _) = expr.kind {
+            // also check for empty `loop {}` statements, skipping those in #[panic_handler]
+            empty_loop::check_empty_loop(cx, expr, block);
+            while_let_loop::check_while_let_loop(cx, expr, block);
+        }
+
+        while_let_on_iterator::check_while_let_on_iterator(cx, expr);
+
+        if let Some((cond, body)) = higher::while_loop(&expr) {
+            while_immutable_condition::check_infinite_loop(cx, cond, body);
+        }
+
+        needless_collect::check_needless_collect(expr, cx);
+    }
+}
+
+fn check_for_loop<'tcx>(
+    cx: &LateContext<'tcx>,
+    pat: &'tcx Pat<'_>,
+    arg: &'tcx Expr<'_>,
+    body: &'tcx Expr<'_>,
+    expr: &'tcx Expr<'_>,
+    span: Span,
+) {
+    let is_manual_memcpy_triggered = manual_memcpy::detect_manual_memcpy(cx, pat, arg, body, expr);
+    if !is_manual_memcpy_triggered {
+        needless_range_loop::check_for_loop_range(cx, pat, arg, body, expr);
+        explicit_counter_loop::check_for_loop_explicit_counter(cx, pat, arg, body, expr);
+    }
+    check_for_loop_arg(cx, pat, arg, expr);
+    for_kv_map::check_for_loop_over_map_kv(cx, pat, arg, body, expr);
+    mut_range_bound::check_for_mut_range_bound(cx, arg, body);
+    single_element_loop::check_for_single_element_loop(cx, pat, arg, body, expr);
+    same_item_push::detect_same_item_push(cx, pat, arg, body, expr);
+    manual_flatten::check_manual_flatten(cx, pat, arg, body, span);
+}
+
+fn check_for_loop_arg(cx: &LateContext<'_>, pat: &Pat<'_>, arg: &Expr<'_>, expr: &Expr<'_>) {
+    let mut next_loop_linted = false; // whether or not ITER_NEXT_LOOP lint was used
+    if let ExprKind::MethodCall(ref method, _, ref args, _) = arg.kind {
+        // just the receiver, no arguments
+        if args.len() == 1 {
+            let method_name = &*method.ident.as_str();
+            // check for looping over x.iter() or x.iter_mut(), could use &x or &mut x
+            if method_name == "iter" || method_name == "iter_mut" {
+                if is_ref_iterable_type(cx, &args[0]) {
+                    explicit_iter_loop::lint_iter_method(cx, args, arg, method_name);
+                }
```

Comment:
I think `check_for_loop_arg` should only decide which function to be called according to the method name. The current implementation of `check_for_loop_arg` seems to do so many works.
And `is_ref_iterable_type` should be called inside `lint_iter_method` because it's a quite detailed implementation of `explicit_iter_loop`.

```suggestion
                explicit_iter_loop::lint_iter_method(cx, args, arg, method_name);
```